### DISCUSSION TITLE
using Depends() to mock function inside endpoint

### DISCUSSION THREAD

**Discussion by jonatasoli** [fastapi/fastapi#6759] [Questions] [answered]
### First check

Hi,

I would like to do a unit test to validate the call from my endpoint and data that will be sent in the body of the request, but I can't replace my subscribe_service function with a mock, would that be possible?

### Example

```Python
@newsletter_router.post(
    "/subscribe",
    response_model=NewsletterResponse,
    status_code=status.HTTP_201_CREATED,
)
async def subscribe_no_auth(
    subscribe_data: NewsletterBase,
    subscribe: subcribe_service = Depends()
):
    try:
        news = await subscribe(subscribe_data)

```
In this case, when I use depends within the function it interprets it as a query param, but I just didn't want it to be a query param it would be there just for me to try to overrides_dependencies with my mock.

Would it be possible?

Thanks

**Comment by myusko** [comment]
Hi @jonatasoli yeah, it will be possible, depends on which library you're using for the testing.

I assume, that you're using `pytest` or default `unittest`, if you're using one of them, please check the next references

[pytest mock](https://github.com/pytest-dev/pytest-mock/)
[unittest mock](https://docs.python.org/3/library/unittest.mock.html)

And four your case, you particularly need to check [unittest mock patch](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) method.

**Comment by jonatasoli** [comment]
Hello @myusko 
I'm using pytest, so when I use pytest-mock it doesn't work in TestClient, because it uses my original function, so I thought about using override_dependences to make the mock, but then I need to use Depends in the function what is making it behave like a query..

**Comment by KiraPC** [comment]
@jonatasoli I faced the same issue and I solved the problem using mocking the function used in the depends.

Let me show you.

```python
@newsletter_router.post("/subscribe")
async def subscribe_no_auth(
    subscribe_data: NewsletterBase,
    subscribe: subcribe_service = Depends(my_function))

def my_function_mock():
    return Mock()

@mock.patch('subscribe_no_auth', new=my_function_mock)
def test_subscribe_no_auth():
     pass
```

**Comment by myusko** [comment]
That's what I have meant, you don't need to mock the `Depends` class, you should mock the function which you are using inside the `Depends`.

**Comment by jonatasoli** [comment]
> @jonatasoli I faced the same issue and I solved the problem using mocking the function used in the depends.
> 
> 
> 
> Let me show you.
> 
> 
> 
> ```python
> 
> @newsletter_router.post("/subscribe")
> 
> async def subscribe_no_auth(
> 
>     subscribe_data: NewsletterBase,
> 
>     subscribe: subcribe_service = Depends(my_function))
> 
> 
> 
> def my_function_mock():
> 
>     return Mock()
> 
> 
> 
> @mock.patch('subscribe_no_auth', new=my_function_mock)
> 
> def test_subscribe_no_auth():
> 
>      pass
> 
> ```

@KiraPC 

i think i'm doing it wrong, when i do so subscribe subscribe_data becomes a mandatory query.

**Comment by jonatasoli** [comment]
> That's what I have meant, you don't need to mock the `Depends` class, you should mock the function which you are using inside the `Depends`.

so when i do that subscrobe_data in addition to focusing on the body also becomes a mandatory query i must be doing something wrong i think.

**Comment by KiraPC** [comment]
@jonatasoli 

I'm not at computer now, anyway, I created a simple boilerplate with fastapi just this morning, and it includes also unittest, you could check it.

https://github.com/KiraPC/fastapi-mvc-boilerplate

Let me know if can be useful for you.

**Comment by jonatasoli** [comment]
@KiraPC 
I made a gist file just to show my current structure, it is similar to what you did, however I am not using classes I am using the functions directly.
In the case of my example it creates a query, in which case would I have to transform my service into a class?
https://gist.github.com/jonatasoli/9864abfd9bc3b4e992981647bdced614

**Comment by KiraPC** [comment, marked as answer]
@jonatasoli  Hi man, I think you have some confusion.

The FastApi Depends function invoke your target function when the endpoint is triggered.

So, in the case of your implementation, defining subscribe as Depends(subscribe_service) means that, when it is evaluated. subscribe is an instance of NewsletterCreateResponse.

But I think that is not what you expect.

So, using a class in this case, can do for you.

```python
class SubscribeService:
     def __init__(self):
           pass

     async def your_magic_fn(self, news_data):
          # do your stuff here

@newsletter_router.post(
    "/subscribe",
    response_model=NewsletterCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
async def subscribe_no_auth(
    subscribe_data: NewsletterBase,
    subscribe: SubscribeService = Depends()
):
    try:
        news = await subscribe.your_magic_fn(subscribe_data)
        return news
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=_detail_msg)
```

I hope now that is clear for you.

**Comment by jonatasoli** [comment]
Hi @KiraPC 

I believe I understand, in my case subscribe_service needs to be inside a class to be able to just invoke the function.
Thanks.

**Comment by KiraPC** [comment]
@jonatasoli 

Yes,

You need to do that because your function need an input parameter.

That is also why your endpoint expect a mandatory query parameter, that for use was NewsLetter model type.

Anyway, does that solved you issue?

**Comment by jonatasoli** [comment]
Hello @KiraPC, yes thanks to explain :D

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @jonatasoli  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
